superset-frontend/src/dashboard/util/getRelatedCharts.ts

```diff
@@ -0,0 +1,112 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Filter } from '@superset-ui/core';
+import { DatasourcesState } from '../types';
+
+type Slice = {
+  datasource: string;
+  slice_id: number;
+};
+
+function getRelatedChartsForSelectFilter(
+  filter: Filter,
+  slices: Record<string, Slice>,
+  datasources: DatasourcesState,
+) {
+  return Object.values(slices)
+    .filter(({ datasource }) => {
+      const chartDatasource = datasources[datasource];
+      if (!chartDatasource) return false;
+      const { column, datasetId } = filter.targets[0] ?? {};
+
+      return (
+        chartDatasource.id === datasetId ||
+        chartDatasource.columns.some(
+          col =>
+            col.column_name === column?.name ||
+            col.column_name === column?.displayName,
+        )
+      );
+    })
+    .map(slice => slice.slice_id);
+}
+function getRelatedChartsForCrossFilter(
+  filterKey: string,
+  slices: Record<string, Slice>,
+  datasources: DatasourcesState,
+): number[] {
+  const sourceSlice = slices[filterKey];
+  if (!sourceSlice) return [];
+
+  const sourceDatasource = datasources[sourceSlice.datasource];
+  if (!sourceDatasource) return [];
+
+  return Object.values(slices)
+    .filter(slice => {
+      if (slice.slice_id === Number(filterKey)) return false;
+      const targetDatasource = datasources[slice.datasource];
+      if (!targetDatasource) return false;
+      if (targetDatasource === sourceDatasource) return true;
+
+      return sourceDatasource.columns.some(sourceColumn =>
+        targetDatasource.columns.some(
+          targetColumn => sourceColumn.column_name === targetColumn.column_name,
```

Comment:
We apply cross filters for charts from different datasets but with the same column name? What ensures that the values are the same? For example, I might have `product` defined in different datasets but the values are completely different.

Reply:
The intent of these changes it to limit the number of unnecessary queries we send to the backend for processing. Right now we are sending any query as long as it is in scope. This is trying to prevent some of those. I believe making sure that filters are actually applicable by comparing the underlying data structures is still responsibility of the backend.

Reply:
> The intent of these changes it to limit the number of unnecessary queries we send to the backend for processing.

I was understanding that we were going to achieve this by limiting the number of unnecessary requests to server. Given that each chart fires a request to server, if they don't belong to the same dataset or are out of scope, then that request would not be fired and the chart would not be loaded.

Reply:
@michael-s-molina I think we still want to allow that columns that belong to different datasets but share the same name might work? I think this was a functionality that also the UI seems to suggest and I don't think we should change it now without revising the functionality more holistically. I believe these changes will already limit a bunch of requests and be an improvement in that sense.

<img width="885" alt="Screenshot 2024-10-03 at 17 29 19" src="https://github.com/user-attachments/assets/9f525070-d576-4399-908a-3aa17fe54fdd">

<img width="558" alt="Screenshot 2024-10-03 at 17 28 57" src="https://github.com/user-attachments/assets/530dac94-5503-4d9a-aebf-0dea1eefd13d">

Reply:
I didn't know this was already allowed. I agree it's important that these changes keep the compatibility.